interpreter.go

```diff
@@ -251,82 +254,117 @@ func addBindings(a, b bindingFrame) bindingFrame {
 	return result
 }
 
-func (i *interpreter) getCurrentEnv(ast ast.Node) environment {
-	return makeEnvironment(
-		i.capture(ast.FreeVariables()),
-		i.stack.getSelfBinding(),
-	)
+func (i *interpreter) newCall(trace TraceElement, env environment, trimmable bool) error {
+	s := &i.stack
+	if s.calls >= s.limit {
+		return makeRuntimeError("max stack frames exceeded.", i.getCurrentStackTrace(trace))
+	}
+	s.newCall(trace, env, trimmable)
+	return nil
 }
 
 func (i *interpreter) evaluate(a ast.Node, tc tailCallStatus) (value, error) {
-	e := &evaluator{
-		trace: &TraceElement{
-			loc:     a.Loc(),
-			context: a.Context(),
-		},
-		i: i,
+	trace := TraceElement{
+		loc:     a.Loc(),
+		context: a.Context(),
 	}
 
 	switch node := a.(type) {
 	case *ast.Array:
 		sb := i.stack.getSelfBinding()
-		var elements []potentialValue
+		var elements []*cachedThunk
 		for _, el := range node.Elements {
-			env := makeEnvironment(i.capture(el.FreeVariables()), sb)
-			elThunk := makeThunk(env, el)
-			elements = append(elements, elThunk)
+			env := makeEnvironment(i.stack.capture(el.FreeVariables()), sb)
+			elThunk := cachedThunk{env: &env, body: el}
+			elements = append(elements, &elThunk)
 		}
 		return makeValueArray(elements), nil
 
 	case *ast.Binary:
-		// Some binary operators are lazy, so thunks are needed in general
-		env := i.getCurrentEnv(node)
-		// TODO(sbarzowski) make sure it displays nicely in stack trace (thunk names etc.)
-		// TODO(sbarzowski) it may make sense not to show a line in stack trace for operators
-		// 					at all in many cases. 1 + 2 + 3 + 4 + error "x" will show 5 lines
-		//					of stack trace now, and it's not that nice.
-		left := makeThunk(env, node.Left)
-		right := makeThunk(env, node.Right)
-
-		builtin := bopBuiltins[node.Op]
-
-		result, err := builtin.function(e, left, right)
-		if err != nil {
-			return nil, err
+		if node.Op == ast.BopAnd {
+			// Special case for shortcut semantics.
+			xv, err := i.evaluate(node.Left, tc)
+			if err != nil {
+				return nil, err
+			}
+			x, err := i.getBoolean(xv, trace)
+			if err != nil {
+				return nil, err
+			}
+			if !x.value {
+				return x, nil
+			}
+			yv, err := i.evaluate(node.Right, tc)
+			if err != nil {
+				return nil, err
+			}
+			return i.getBoolean(yv, trace)
+		} else if node.Op == ast.BopOr {
+			// Special case for shortcut semantics.
+			xv, err := i.evaluate(node.Left, tc)
+			if err != nil {
+				return nil, err
+			}
+			x, err := i.getBoolean(xv, trace)
+			if err != nil {
+				return nil, err
+			}
+			if x.value {
+				return x, nil
+			}
+			yv, err := i.evaluate(node.Right, tc)
+			if err != nil {
+				return nil, err
+			}
+			return i.getBoolean(yv, trace)
+
+		} else {
+			left, err := i.evaluate(node.Left, tc)
+			if err != nil {
+				return nil, err
+			}
+			right, err := i.evaluate(node.Right, tc)
+			if err != nil {
+				return nil, err
+			}
+			builtin := bopBuiltins[node.Op]
```

Comment:
You're talking about saving a dereference, right?  I wonder if we could just make bopBuiltins map to actual structs (rather than pointers). In fact, bopBuiltins is not even an array, it's a map from enum. I'm not sure if that's optimized to an array (probably not).  Maybe a switch is better.  I'll add a TODO.

Reply:
Scratch that, it is an array not a map.  The type says so.  It's just initialized using the map syntax.